Add SystemUtil for querying common system properties

clintval opened the pull request:
Functionality pulled from [`fgbio`](https://github.com/fulcrumgenomics/fgbio/blob/8f9d01ab4bc4b61e8e7ac696c5e2a91e3165f705/src/main/scala/com/fulcrumgenomics/cmdline/FgBioMain.scala#L46) and [`dagr`](https://github.com/fulcrumgenomics/dagr/pull/368).

Which of these would you like as a part of public API?

clintval committed:
Add SystemUtil for querying common system properties

nh13 commented:
@clintval what do you think about these changes? https://github.com/clintval/commons/pull/1.  I'm happy to update fgbio and dagr for you.

clintval commented:
A+ go for it.

Was going to write tests after the public API was revealed, but thanks for beating me to it!

clintval committed:
Fix cross-Scala compatibility with Regex.matches()

clintval commented on src/main/scala/com/fulcrumgenomics/commons/util/SystemUtil.scala:
```diff
@@ -0,0 +1,79 @@
+  def isOsVersion(name: String): Boolean = OsVersion.contains(name)
+
+  /** Returns true if the operating system version matches the given regular expression, false otherwise. */
+  def isOsVersion(regex: Regex): Boolean = OsVersion.exists(v => regex.findFirstMatchIn(v).isDefined)
```
The member `Regex.matches()` does not exist in Scala 2.12, so I switched to `findFirstMatchIn()`.

But, that is not correct because we want full match anchoring!

For example. This is now possible, but shouldn't be:

```scala
scala> OsVersion
Some(0...0.00.0..0.10.14.6)
scala> isOsVersion("10\\.14\\.".r)
true
```

I'll try again.

clintval committed:
Use 2.12 and 2.13 compatible regex anchor matching

clintval commented:
@nh13, all set. I found a way to have anchored regex matching that is 2.12 and 2.13 compatible. Ping me if you want help with the rest.

clintval committed:
Fix regex to respect an anchored match

clintval commented on src/main/scala/com/fulcrumgenomics/commons/util/SystemUtil.scala:
```diff
@@ -0,0 +1,79 @@
+  lazy val IsOsLinux: Boolean = LinuxNamePrefixes.exists(prefix => OsName.exists(_.startsWith(prefix)))
+
+  /** True if this operating system is Mac, false otherwise. */
+  lazy val IsOsMac: Boolean = MacNamePrefixes.exists(prefix => OsName.exists(_.startsWith(prefix)))
```
Would you rather us just use the Scala builtins?

https://www.scala-lang.org/api/current/scala/util/Properties$.html#isMac:Boolean

nh13 replied:
That makes a lot more sense, nice pickup

clintval reviewed (commented):
I would be interested in seeing this merged as I want to factor this into `PicardTask` in Dagr.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/util/SystemUtil.scala:
```diff
@@ -0,0 +1,79 @@
+
+  /** Gets a system property.  Returns None if not found or we are not allowed to retrieve the property. */
+  def getSystemProperty(property: String): Option[String] = {
+    try { Option(System.getProperty(property)) }
```
Could use builtin? https://www.scala-lang.org/api/current/scala/util/Properties$.html#isMac:Boolean

nh13 reviewed (changes requested):
One suggested change you yourself suggested.

clintval committed:
Fallback to implementations in scala.util.Properties

clintval commented:
Few changes:

- Copyrights to 2020
- Changed `getSystemProperty` to `safePropOrNone` to match the ecosystem of Scala property getters in [`scala.util.Properties`](https://www.scala-lang.org/api/current/scala/util/Properties$.html#propIsSet(name:String):Boolean) for when you want to get a system property safely (catch security manager exceptions)
- Deleted `lazy val OsName` as that's in the [Scala builtins](https://www.scala-lang.org/api/current/scala/util/Properties$.html#osName:String)
- Moved over implementations to other Scala builtins like `isLinux`, `isMac`, and `propOrNone`

clintval commented:
@nh13 I think this is merge-ready FYI

clintval commented:
If possible, I'd like to find a way to move this along because I also want to use these features for patches to `fgbio` and `dagr`. I'm still having difficulty with the occasional core dump on MacOS (`10.15.3`) due to the Intel infalters/deflaters and want to unblock https://github.com/fulcrumgenomics/dagr/pull/368.
